spec/app/ics-721-nft-transfer/README.md

```diff
@@ -0,0 +1,386 @@
+
+## Synopsis
+
+This standard document specifies packet data structure, state machine handling logic, and encoding details for the transfer of non-fungible tokens over an IBC channel between two modules on separate chains. The state machine logic presented allows for safe multi-chain `classId` handling with permissionless channel opening. This logic constitutes a "non-fungible token transfer bridge module", interfacing between the IBC routing module and an existing asset tracking module on the host state machine, which could be either a Cosmos-style "native" module or a smart contract running in a virtual machine.
+
+### Motivation
+
+Users of a set of chains connected over the IBC protocol might wish to utilise an asset issued on one chain on another chain, perhaps to make use of additional features such as exchange or privacy protection, while retaining uniqueness with the original asset on the issuing chain. This application-layer standard describes a protocol for transferring non-fungible tokens between chains connected with IBC which preserves asset uniqueness, preserves asset ownership, limits the impact of Byzantine faults, and requires no additional permissioning.
+
+### Definitions
+
+The IBC handler interface & IBC routing module interface are as defined in [ICS 25](../../core/ics-025-handler-interface) and [ICS 26](../../core/ics-026-routing-module), respectively.
+
+### Desired Properties
+
+- Preservation of uniqueness (two-way peg).
+- Preservation of total supply (maintained on a single source chain & module).
+- Permissionless token transfers, no need to whitelist connections, modules, or `classId`s.
+- Symmetric (all chains implement the same logic, no in-protocol differentiation of hubs & zones).
+- Fault containment: prevents Byzantine-inflation of tokens originating on chain `A`, as a result of chain `B`'s Byzantine behaviour (though any users who sent tokens to chain `B` may be at risk).
+
+## Technical Specification
+
+### Data Structures
+
+Only one packet data type is required: `NonFungibleTokenPacketData`, which specifies the class id, class uri, token id's, token uri's, sending account, and receiving account.
+
+```typescript
+interface NonFungibleTokenPacketData {
+  classId: string
+  classUri: string
+  tokenIds: []string
+  tokenUris: []string
+  sender: string
+  receiver: string
+}
+```
+`classId` uniquely identifies the class/collection to which this NFT belongs in the originating host environment.  In the case of an ERC-1155 compliant smart contract, for example, this could be a string representation of the top 128 bits of the token ID.
+
+`classUri` is optional, but will be extremely beneficial for cross-chain interoperability with NFT marketplaces like OpenSea, where [class/collection metadata](https://docs.opensea.io/docs/contract-level-metadata) can be added for better user experience.
```

Comment:
why exactly do we want to pass this along as on-chain data, if it's just a pointer to off-chain data anyways? can't anyone interested in the off-chain data look up the source chain, look up the URI there, and then look up the data? this introduces the possibility of divergence (e.g. a URI on some receiving chain not matching the URI associated with the asset on the original source chain), which seems like a recipe for problems

Reply:
Is it possible for the classUri to change after initialization? Similar question regarding tokenUris?

> Avoiding metadata divergence is something that can and should be enforced at the application layer.

This can be done at the point of transfer, but if the metadata changes at a later date, it may be difficult to propogate this information over to wherever the NFT is currently

Reply:
In ERC721 URIs may be changed after minting if an account is authorised to make that change (can also be locked on or after mint).

If we were to use a middleware approach and someone opted to clone metadata, would want to pass the origin clone time.

Reply:
One way to enforce the immutability of NFT metadata would be to update the implementation in [x/nft](https://github.com/cosmos/cosmos-sdk/blob/master/x/nft/keeper/nft.go) to prevent anyone from updating the metadata (i.e., `uri` and `uri_hash`) of **IBC tokens** they own.

We shouldn't worry about IBC tokens locked in (owned by) escrow accounts along the IBC path because no one can update them anyway.

Reply:
> Is it possible for the classUri to change after initialization? Similar question regarding tokenUris?
> 
> > Avoiding metadata divergence is something that can and should be enforced at the application layer.
> 
> This can be done at the point of transfer, but if the metadata changes at a later date, it may be difficult to propogate this information over to wherever the NFT is currently

It's possible that the `classUri` and other metadata (e.g., name, symbol) of an NFT class could change after associated NFTs have been transferred out via IBC.  However, this is not a new thing, we have the same "problem" with fungible tokens, as the metadata (name, symbol etc.) of a fungible token can change at any time.  Such issues are being resolved out-of-band by UI teams and the same should be done with NFT classes.

Reply:
I think it's too limited for a general usage, if all dapps will need to freeze a token metadata when IBCing it. I think adding a timestamp makes most of sense, and clarifying on the protocol side - that it should be taken with consideration that it's not a guaranteed by the standard that the token metadata (in particular the URI) won't change. 
So that would be a not required , maybe redundant information, but maybe useful for some use cases.

Reply:
TBH, ideally we would like to use DID here, an reference the NFT with DID. I was talking already about it during the initial NFT proposal last summer. For functioning we would need a resolver, but this is not needed to ship the NFT standard, and that functionality could be added later without changing the standard or the base protocol.

Reply:
> I think it's too limited for a general usage, if all dapps will need to freeze a token metadata when IBCing it. I think adding a timestamp makes most of sense, and clarifying on the protocol side - that it should be taken with consideration that it's not a guaranteed by the standard that the token metadata (in particular the URI) won't change. So that would be a not required , maybe redundant information, but maybe useful for some use cases.

Essentially, the NFT metadata -- a piece of descriptive information stored off-chain -- tells dApps what the NFT looks like (image/logo) and what traits (properties) are visually presentable to human users.

For an NFT that represents digital artwork, its metadata is all that matters. For other more "business-heavy" scenarios where, for example, an NFT represents digital security, the token's on-chain **data** holds the security's business state and matters more than its metadata.

Now, let's look at why people want to transfer their NFTs to other chains. Nowadays in the domain of public blockchains, most if not all of the motivation is to transfer tokens to a popular NFT marketplace for trading, in which case NFT metadata in a format that is compatible with the target marketplace (OpenSea being the dominant standard today) is of utmost importance.

What we had in mind, maybe unconsciously, for this ICS721 spec is just for that purpose -- sending tokens to popular NFT marketplace for trading, that's why our IBC packet only carries along `tokenUri` (pointing to metadata) but not NFT data itself.

If we were to send NFTs to another chain to take advantage of its enhanced on-chain capabilities that are not available on the source chain, then we must pass along NFT data in our IBC packets as well (of course, the NFT data must be in a format the target chain can understand). For example, we may want to send our digital security tokens to a chain that support security mortgaging.

IMHO, in the context of NFT trading, metadata represents the unique value people trade for and therefore should be change along the IBC path.

Reply:
I agree. That's why I propose to include the timestamp ( the time of tracked the last modification) as well. Or hash of the whole onchain object.

Reply:
Thanks for the comment. As you said, this timestamp/hash is not required and may be redundant; the cost and trouble involved -- adding a field in NFT object (updating `x/nft`) and adding a field in the IBC packet -- most likely outweighs the benefits.

Updating IBC token metadata (`tokenUri`) by some sink chain is done through a local tx, so there is really no need for a separate timestamp as the information is already available in the on-chain tx. (redundant)

On the other hand, even if we pass the timestamp of latest metadata update back to the source chain in the IBC packet, we can't do much about it because there is no way for us to verify, in the protocol, whether the off-chain resource referenced by the updated uri is compatible/usable by the source-chain dApps; therefore, we can't and shouldn't override `tokenUri` in the un-escrowed token.

So, maybe it's safe to skip this timestamp/hash field for now, you think?

Reply:
x/nft is not yet released, so there is no much cost to add additional field. I would like to hear more opinions if there is a benefit from the business side for including timestamp or hash. PS: for CID links, hash is not needed, so that could be a workaround.

Reply:
I believe this is the last issue to be resolved for this spec: do we enforce the freezing of NFT token `uri` in ICS-721 spec? @AdityaSripal

Reply:
Maybe we don't need to try to answer this question in the specification, but I still think it's a recipe for ecosystem-level disaster if NFT infrastructure ends up not checking the source chain. Even if there is some option to freeze the metadata here, the infrastructure then needs to check that every intermediate chain implemented this ICS correctly and didn't malleate the metadata... of course, there is a similar concern for ICS20 multi-hop transfer - but the infrastructure should be checking there too!

I don't want to block too long on this concern, however. Could we add a paragraph writeup explaining the risks involved with using this field?

Reply:
Agreed with Chris, that for now we can just write a paragraph that users are responsible for verifying metadata at the source chain